### PR TITLE
Arbitrary self types v2: pointers feature gate.

### PR DESCRIPTION
**Pull request by adetaylor**
The main `arbitrary_self_types` feature gate will shortly be reused for a new version of arbitrary self types which we are amending per [this RFC](https://github.com/rust-lang/rfcs/blob/master/text/3519-arbitrary-self-types-v2.md). The main amendments are:

* _do_ support `self` types which can't safely implement `Deref`
* do _not_ support generic `self` types
* do _not_ support raw pointers as `self` types.

This PR relates to the last of those bullet points: this strips pointer support from the current `arbitrary_self_types` feature. We expect this to cause some amount of breakage for crates using this unstable feature to allow raw pointer self types. If that's the case, we want to know about it, and we want crate authors to know of the upcoming changes.

For now, this can be resolved by adding the new
`arbitrary_self_types_pointers` feature to such crates. If we determine that use of raw pointers as self types is common, then we may maintain that as an unstable feature even if we come to stabilize the rest of the `arbitrary_self_types` support in future. If we don't hear that this PR is causing breakage, then perhaps we don't need it at all, even behind an unstable feature gate.

[Tracking issue](https://github.com/rust-lang/rust/issues/44874)

This is [step 4 of the plan outlined here](https://github.com/rust-lang/rust/issues/44874#issuecomment-2122179688)

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs [hunk 1]
```diff
@@ -1651,6 +1651,14 @@ fn check_fn_or_method<'tcx>(
     }
 }
 
+/// The `arbitrary_self_types_pointers` feature implies `arbitrary_self_types`.
+#[derive(Clone, Copy)]
```
Comment (wesleywiser) [inline compiler/rustc_hir_analysis/src/check/wfcheck.rs:1655]: I would derive at least `PartialEq` and then you can avoid a bunch of `matches!(foo, ArbitrarySelfTypesLevel::...)` elsewhere.

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs [hunk 2]
```diff
@@ -1651,6 +1651,14 @@ fn check_fn_or_method<'tcx>(
...
+enum ArbitrarySelfTypesLevel {
+    None,                       // neither arbitrary_self_types nor arbitrary_self_types_pointers
```
Comment (wesleywiser) [inline compiler/rustc_hir_analysis/src/check/wfcheck.rs:1657]: Nit: To me, this is often a sign you should remove this variant and use `Option<ArbirarySelfTypesLevel>` in the codebase. If you find yourself wanting the combinators that come with `Option`, consider refactoring this.

**Review by wesleywiser — APPROVED** [review]
Just a few minor suggestions 🙂

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Looking at results for `arbitrary_self_types` on GH search, it seems the majority of hits are from forks of `PyO3` (which mentions the feature in a comment, does not currently use it) and forks of `rustc`. Therefore, I don't think we need to worry too much about trying to migrate current users of the unstable feature to (potentially) the newly introduced pointer feature gate. 

If you concur, I'm happy to r+!

**Comment by adetaylor** [comment]
I do concur, but also, (unless it's vastly more expensive than I expect) - could we do a crater run? No harm in catching unexpected surprises!

**Comment by adetaylor** [comment]
OK, I've been told that it's pretty expensive to do a crater run and not all that useful for nightly-only breakage, so I'll take the r+ then!

**Comment by wesleywiser** [comment]
We're right at the beginning of the release cycle so if there is unexpected, widespread breakage we have plenty of time to fix or revert before beta branches so I think this is the appropriate time to:

@bors r+

**Comment by adetaylor** [comment]
This didn't work as planned - I'm seeing an [unexpected error in one of my crates which relies upon the nightly `arbitrary_self_types` feature](https://github.com/google/autocxx/actions/runs/10731172174/job/29760831324). I'm going to look into why, but we might need to revert this if a fix isn't immediately apparent. Looking into it now.

**Comment by purplesyringa** [comment]
This broke [crossmist](https://crates.io/crates/crossmist). crossmist needs to serialize and deserialize objects all the time, including `Box<dyn Trait>`, sort of like [serde_traitobject](https://crates.io/crates/serde_traitobject) (after a quick look, that crait likely broke too).

While serializing through `dyn Trait` is easy if the trait contains something like a `serialize_me` method, deserializing requires somehow creating a valid fat pointer to call a method on from thin air. Both crossmist and serde_traitobject achieve this by storing a vtable pointer prior to object data during serialization, and invoke the virtual method on a dangling fat pointer (with the right vtable) during deserialization.

Crucially, this is only valid if pointers are used, not references, because references can't be dangling or null.

There are ways to workaround this, e.g. by using ZSTs or adding a `fn get_deserializing_fn(&self) -> fn(&mut Deserializer) -> Self` function to the trait and prepending the "deserializing fn" to data instead of the vtable. However, all these workarounds introduce indirection and slow code down when really the expected behavior is quite simple.

**Comment by adetaylor** [comment]
@purplesyringa thanks for the report, good to know. First off, from my side, there's no particular pressure to _remove_ `arbitrary_self_types_pointers`, it's just that we're hoping to eventually stabilize _the rest of_ arbitrary self types, so that's why it's been split into a separate unstable feature gate. Others may have different views, but that's my current understanding.

Maybe your use-case could be accommodated using "regular" `arbitrary_self_types` by transmuting your fat pointer into a transparent newtype wrapper which implements the upcoming `Receiver` trait, and then using that as the receiver? But maybe not if you're trying to call existing functions on existing types, and don't have an opportunity to insert an additional type wrapper.

**Comment by purplesyringa** [comment]
Sure, I just saw this in the OP and thought I should weigh in to prevent this outcome:

> If we don't hear that this PR is causing breakage, then perhaps we don't need it at all, even behind an unstable feature gate.

---

Transmuting the fat pointer into a newtype wrapper is fine, the more troublesome part is object safety. The method `fn deserialize_me(self: NewTypeWrapper<Self>)` would have to be callable via `Self as Deserialize`'s vtable. This interacts with `DispatchFromDyn`, which I'm unfamiliar with, and it's not 100% clear to me how stable it is to stabilization and if I can easily wrap a raw pointer using it.

**Comment by adetaylor** [comment]
`DispatchFromDyn` is on its way to partial stabilization via the [`#[derive(SmartPointer)]` work](https://rust-lang.github.io/rfcs/3621-derive-smart-pointer.html) in case that works for you?

**Comment by purplesyringa** [comment]
It might work, thanks!

**Pull request merged**